LocalDateTime.dfy

```
@@ -1,113 +1,371 @@
+        var day := ToInt(dayStr);
+        var hour := ToInt(hourStr);
+        var minute := ToInt(minuteStr);
+        var second := ToInt(secondStr);
+        var millisecond := ToInt(millisecondStr);
+
+        Of(year, month, day, hour, minute, second, millisecond)
+  }
 
   // Formatting functions
+  // ISO 8601 format
   function ToString(dt: LocalDateTime): string
     requires IsValidLocalDateTime(dt)
+  {
+    OfInt(dt.year) + "-" +
+    DTUtils.PadWithZeros(dt.month, 2) + "-" +
+    DTUtils.PadWithZeros(dt.day, 2) + "T" +
+    DTUtils.PadWithZeros(dt.hour, 2) + ":" +
+    DTUtils.PadWithZeros(dt.minute, 2) + ":" +
+    DTUtils.PadWithZeros(dt.second, 2) + "." +
+    DTUtils.PadWithZeros(dt.millisecond, 3)
+  }
 
   function Format(dt: LocalDateTime, pattern: string): string
     requires IsValidLocalDateTime(dt)
-
-  // External dependencies (to be implemented per target language)
-  function {:extern} NowImpl(): LocalDateTime
-    ensures IsValidLocalDateTime(NowImpl())
+  {
+    if pattern == "yyyy-MM-dd" then
+      OfInt(dt.year) + "-" + DTUtils.PadWithZeros(dt.month, 2) + "-" + DTUtils.PadWithZeros(dt.day, 2)
+    else if pattern == "HH:mm:ss" then
+      DTUtils.PadWithZeros(dt.hour, 2) + ":" + DTUtils.PadWithZeros(dt.minute, 2) + ":" + DTUtils.PadWithZeros(dt.second, 2)
+    else if pattern == "yyyy-MM-dd HH:mm:ss" then
+      OfInt(dt.year) + "-" + DTUtils.PadWithZeros(dt.month, 2) + "-" + DTUtils.PadWithZeros(dt.day, 2) + " " +
+      DTUtils.PadWithZeros(dt.hour, 2) + ":" + DTUtils.PadWithZeros(dt.minute, 2) + ":" + DTUtils.PadWithZeros(dt.second, 2)
+    else if pattern == "dd/MM/yyyy" then
+      DTUtils.PadWithZeros(dt.day, 2) + "/" + DTUtils.PadWithZeros(dt.month, 2) + "/" + OfInt(dt.year)
+    else if pattern == "MM/dd/yyyy" then
+      DTUtils.PadWithZeros(dt.month, 2) + "/" + DTUtils.PadWithZeros(dt.day, 2) + "/" + OfInt(dt.year)
+    else
+      // Default to ISO 8601 format, yyyy-MM-ddTHH:mm:ss.fff
+      ToString(dt)
```

This isn't a good experience when a caller has a typo - if you pass "yyyy/MM/dd" for example, you don't get an error, you get the wrong result (from the caller's perspective). 

It would be better to at least return a Failure on a format you don't support at least. Even better, though, I'd suggest defining a `datatype` of the formats you support and accept that as a parameter instead of a string, so the type system ensures you can't pass anything else in.